src/libopensc/card-jpki.c

```diff
@@ -0,0 +1,344 @@
+}
+
+static int
+jpki_read_binary(sc_card_t * card, unsigned int idx,
+				 u8 * buf, size_t count, unsigned long flags)
+{
+	struct sc_card_driver *iso_drv = sc_get_iso7816_driver();
+	const struct sc_card_operations *iso_ops = iso_drv->ops;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	return iso_ops->read_binary(card, idx, buf, count, flags);
+}
+
+static int
+jpki_pin_cmd(sc_card_t * card, struct sc_pin_cmd_data *data, int *tries_left)
+{
+	int rc;
+	sc_path_t path;
+	sc_apdu_t apdu;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (tries_left) {
+		*tries_left = -1;
+	}
+
+	switch (data->pin_reference) {
+	case 1:
+		sc_format_path("0018", &path);
+		path.type = SC_PATH_TYPE_FILE_ID;
+		rc = sc_select_file(card, &path, NULL);
+		break;
+	case 2:
+		sc_format_path("001b", &path);
+		path.type = SC_PATH_TYPE_FILE_ID;
+		rc = sc_select_file(card, &path, NULL);
+		break;
+	default:
+		sc_log(card->ctx, "Unknown PIN reference: %d", data->pin_reference);
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	LOG_TEST_RET(card->ctx, rc, "SELECT_FILE error");
+
+	switch (data->cmd) {
+	case SC_PIN_CMD_VERIFY:
+		sc_format_apdu(card, &apdu, SC_APDU_CASE_3, 0x20, 0x00, 0x80);
+		apdu.data = data->pin1.data;
+		apdu.datalen = data->pin1.len;
+		apdu.lc = data->pin1.len;
+		rc = sc_transmit_apdu(card, &apdu);
+		LOG_TEST_RET(card->ctx, rc, "APDU transmit failed");
+		break;
+	case SC_PIN_CMD_GET_INFO:
+		sc_format_apdu(card, &apdu, SC_APDU_CASE_1, 0x20, 0x00, 0x80);
+		rc = sc_transmit_apdu(card, &apdu);
+		LOG_TEST_RET(card->ctx, rc, "APDU transmit failed");
+		if (apdu.sw1 != 0x63) {
+			sc_log(card->ctx, "VERIFY GET_INFO error");
+			return SC_ERROR_CARD_CMD_FAILED;
+		}
+		if (tries_left) {
+			*tries_left = apdu.sw2 - 0xC0;
+			sc_log(card->ctx, "tries_left=%d\n", *tries_left);
+		}
+		break;
+	default:
+		sc_log(card->ctx, "Card does not support PIN command: %d", data->cmd);
+		return SC_ERROR_NOT_SUPPORTED;
+	}
+
+	return SC_SUCCESS;
+}
+
+static int
+jpki_set_security_env(sc_card_t * card,
+					  const sc_security_env_t * env, int se_num)
+{
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL,
+			 "flags=%08x op=%d alg=%d algf=%08x algr=%08x kr0=%02x, krfl=%d\n",
+			 env->flags, env->operation, env->algorithm,
+			 env->algorithm_flags, env->algorithm_ref, env->key_ref[0],
+			 env->key_ref_len);
+	switch (env->operation) {
+	case SC_SEC_OPERATION_SIGN:
+		break;
+	default:
+		return SC_ERROR_NOT_SUPPORTED;
+	}
+	return SC_SUCCESS;
+}
+
+static int
+jpki_compute_signature(sc_card_t * card,
+					   const u8 * data, size_t datalen, u8 * out, size_t outlen)
+{
+	int rc;
+	sc_path_t path;
+	sc_apdu_t apdu;
+	unsigned char resp[SC_MAX_APDU_BUFFER_SIZE];
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	sc_format_path("0017", &path);
+	path.type = SC_PATH_TYPE_FILE_ID;
+	rc = sc_select_file(card, &path, NULL);
+	SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, rc, "select failed");
```

Comment:
Exactly, This is selecting sign key as IEF.
moved to jpki_set_security_env(). adf9904